plugins/qCompass/ccTrace.cpp

```diff
@@ -0,0 +1,678 @@
+	CCVector3 QS(segStart->x - query->x, segStart->y - query->y, segStart->z - query->z);
+	CCVector3 QE(segEnd->x - query->x, segEnd->y - query->y, segEnd->z - query->z);
+	
+	//is angle between these vectors obtuce (i.e. QS dot QE) < 0)? If so we are inside a circle between start&end, otherwise we are not
+	QS.normalize();QE.normalize();
+	float dot = QS.dot(QE);
+	return QS.dot(QE) < 0;
+}
+
+bool ccTrace::optimizePath(int maxIterations)
+{
+	bool success = true;
+
+	if (m_waypoints.size() < 2)
+	{
+		m_trace.clear();
+		return false; //no segments...
+	}
+
+	#ifdef DEBUG_PATH
+	int idx = m_cloud->getScalarFieldIndexByName("Search"); //look for scalar field to write search to
+	if (idx == -1) //doesn't exist - create
+		idx=m_cloud->addScalarField("Search");
+	m_cloud->setCurrentScalarField(idx);
+	#endif
+	
+	//loop through segments and build/rebuild trace
+	int start, end, tID; //declare vars
+	for (unsigned i = 1; i < m_waypoints.size(); i++)
+	{
+		//calculate indices
+		start = m_waypoints[i - 1]; //global point id for the start waypoint
+		end = m_waypoints[i]; //global point id for the end waypoint
+		tID = i - 1; //id of the trace segment id (in m_trace vector)
+
+		//are we adding to the end of the trace?
+		if (tID >= m_trace.size()) 
+		{
+			std::deque<int> segment = optimizeSegment(start, end, m_search_r, maxIterations); //calculate segment
+			m_trace.push_back(segment); //store segment
+			success = success && !segment.empty(); //if the queue is empty, we failed
+		} else //no... we're somewhere in the middle - update segment if necessary
+		{
+			if (!m_trace[tID].empty() && (m_trace[tID][0] == start) &&  (m_trace[tID][m_trace[tID].size() - 1] == end)) //valid trace and start/end match
+				continue; //this trace has already been calculated - we can skip! :)
+			else
+			{
+				//calculate segment
+				std::deque<int> segment = optimizeSegment(start, end, m_search_r, maxIterations); //calculate segment
+				success = success && !segment.empty(); //if the queue is empty, we failed
+
+				//add trace
+				if (m_trace[tID][m_trace[tID].size() - 1] == end) //end matches - we can replace the current trace & things should be sweet (all prior traces will have been updated already)
+					m_trace[tID] = segment; //end is correct - overwrite this block, then hopefully we will match in the next one
+				else //end doesn't match - we need to insert
+					m_trace.insert(m_trace.begin()+tID, segment);
+			}
+		}
+	}
+
+	#ifdef DEBUG_PATH
+	CCLib::ScalarField * f = m_cloud->getScalarField(idx);
+	f->computeMinAndMax();
+	#endif
+
+	return success;
+}
+
+int ccTrace::COST_MODE = ccTrace::MODE::DARK; //set default cost mode
+std::deque<int> ccTrace::optimizeSegment(int start, int end, float search_r, int maxIterations, int offset)
+{
+	//check handle to point cloud
+	if (!m_cloud) return std::deque<int>(); //error -> no cloud
+
+	//retreive and store start & end rgb
+	const ColorCompType* s = m_cloud->getPointColor(start);
```

Comment:
here you must check if the cloud has colors, if not getPointColor causes a segfault